### PR TITLE
SINGA -475 add Div operator implementation to singa

### PR DESCRIPTION
**Pull request by pinpom**

### CODE REVIEW

File: python/singa/autograd.py [hunk 1]
```diff
@@ -1787,6 +1787,22 @@ def abs(a):
...
+    def _init_(self):
+        super(Div, self)._init_()    
+    
+    def forward(self, a, b):    
+        if training:
+            self.input = (a, b)
+            return singa.__div__(a, b)
+
+    def backward(self, dy):
+        return singa.__div__(dy, self.input[0]), singa.__mul__(dy, self.input[1])
```
Comment (joddiy) [inline python/singa/autograd.py]: the gradients should be:
dy/dx_1 = b^-1
dy/dx_2 = -a*b^-2

File: test/python/test_operation.py [hunk 2]
```diff
@@ -322,6 +322,21 @@ def test_LeakyRelu(self):
...
+        y = tensor.from_numpy(Y)
+        x.to_device(gpu_dev) 
+        y.to_device(gpu_dev) 
+
+        result=autograd.div(x,y)
+        dx=result.creator.backward(x.data)[0]  
+
+        result_np=np.divide(X,Y)      
+        np.testing.assert_array_almost_equal(tensor.to_numpy(result), result_np)
+        self.check_shape(dx.shape(), (3, 2))
```
Comment (joddiy) [inline test/python/test_operation.py]: please add the test case of backward propagation.

Comment (pinpom) [reply test/python/test_operation.py]: fixed backward function and added test

File: test/python/test_operation.py [hunk 3]
```diff
@@ -34,6 +34,35 @@
...
+        while not it.finished:
+            iy = it.multi_index
+            old_value = y[iy]
+            y[iy] = old_value + h # increment by h
+            fyh = f(x,y) # evaluate f(y + h)
+            y[iy] = old_value # restore to previous value (very important!) 
+            grad[iy] = (fyh - fx) / h
+            it.iternext()
+        return grad
+
```
Comment (joddiy) [inline test/python/test_operation.py]: can we optimize this part as: 
```
def eval_numerical_gradient_b(f, x, y, reverse = False):
    h = 0.00001
    grad = np.zeros(x.shape)
    t = y if reverse else x
    fx = f(x, y)
    it = np.nditer(t, flags=['multi_index'], op_flags=['readwrite'])
    while not it.finished:
        _it = it.multi_index
        old_value = t[_it]
        t[_it] = old_value + h # increment by h
        fth = f(x, y) # evaluate f(x + h)
        t[_it] = old_value # restore to previous value (very important!) 
        grad[_it] = (fth - fx) / h # the slope
        it.iternext() # step to next dimension
    return grad
```

Comment (pinpom) [reply test/python/test_operation.py]: yes. fixed the code accordingly

**Comment by pinpom** [comment]
fixed the backward test function to calculating specific derivative

**Comment by chrishkchris** [comment]
I see the travis-ci error log is like this
`fatal: unable to access 'https://github.com/apache/incubator-singa.git/': Failed to connect to github.com port 443: Connection timed out`
Maybe the server is down? maybe need to wait

**Comment by chrishkchris** [comment]
This PR is closed. A new PR is opened to replace this PR.

**Pull request closed without merging**